### ISSUE TITLE
Macros KS_EXTRA keycodes change file encoding

### ISSUE THREAD

**Issue by iddqd505** [vim/vim#13916]
### Steps to reproduce

I noticed when I record a simple macro with some combinations of modifier keys "Control" or "Alt" or make a motion like f) and then paste that macro register content in a buffer Vim prints some strange characters I think are called "KS_EXTRA keycodes"

1. I record a macro for this test line 
`TESTLINE_2001 (testing_test)`

2. Macro w (qw)
`0<C-A>f)vi)<C-C><M-j><M-j>`

3. (PASTE 1) This is the content of that macro pasted with "wp
`0f)ýavi)üCüjüj`

4. (PASTE 2) This is the same macro w pasted with execute command :echo keytrans (@w)
`0<C-A>f)<t_<fd>a>vi)<C-C><M-j><M-j>`

5. (PASTE 3) And this is the same macro w pasted in insert mode with Ctrl-R+w
`0f)<t_ý>avi)üCjj`

6. Aside getting 3 different pastes the real problem is when I leave written the macro register then save and close the file this happens:

    6.1. The file encoding changes from utf-8 to none

    6.2. Sometimes the buffer opens like read only file

    6.3. The pasted macro register changes from original form (PASTE 1) to (PASTE 3)

7. As a side note I have just noticed if I copy-paste (Ctrl+C and Ctrl+V in Windows) the macro register content (PASTE 1) the result also changes to (PASTE 3)

### Expected behaviour

Im not a programmer but I use gVim on Windows 10 as my primary text editor for a long time and the macros contents never pasted these extra keycodes until now

Also all my text files are in utf-8 encoding and unix fileformat since it helps with compatibility with Linux systems and gVim never had a problem to edit them that way until now

Im not an advanced Vim user and when I try to look for references or help about this problem I always get entangled in the rabbit hole that is file encoding configurations

I also have not done more extensive testing because like I say I actually have no idea what am I doing in this particular field

Please help

### Version of Vim

gVim version 9.0.2189

### Environment

Windows 10 Version 21H2 (OS Build 19044.1288)

Here is the output of the :version command:
https://github.com/iddqd505/dotfiles/blob/main/vim%20version

Here is my :scriptnames list of plugins. However I still can replicate the problem exactly the same even if I remove all my installed plugins and reload gVim
https://github.com/iddqd505/dotfiles/blob/main/vim%20scriptnames

And here is a sanitized and debloated version of my _vimrc file:
https://github.com/iddqd505/dotfiles/blob/main/_vimrc

Edited my post to improve formatting and also added sample files along with sample screenshots. Here are the links

sample file 01 before
https://github.com/iddqd505/dotfiles/blob/main/sample%20file%2001%20before

sample file 01 before.jpg
https://github.com/iddqd505/dotfiles/blob/main/sample%20file%2001%20before.jpg

sample file 02 after
https://github.com/iddqd505/dotfiles/blob/main/sample%20file%2002%20after

sample file 02 after.jpg
https://github.com/iddqd505/dotfiles/blob/main/sample%20file%2002%20after.jpg

### Logs and stack traces

Updated to Vim version 9.1.0061
Downloaded from
https://github.com/vim/vim-win32-installer/releases/tag/v9.1.0061
gvim_9.1.0061_x64_signed.exe

Uploaded screenshot with new error results:
https://github.com/iddqd505/dotfiles/blob/main/Sample%20file%2003%20before-and-after%20in%20V9_1_0061.png

**Comment by zeertzjq** [comment]
Maybe its necessary to add an actual notation for `K_NOP`?

**Comment by chrisbra** [comment]
can you please fix the formatting of your how-to-reproduce section? I have troubles following the guide. Please use markdown code formatting if possible

**Comment by iddqd505** [comment]
Yes sorry for the late response since I dont know Github I also had trouble trying to fit the how to steps in the required post format but I tried to follow the posting rules

I will try to fix it also I have uploaded 2 sample files with a before and after I open them with my gVim editor alongside screenshots of how they show in my gVim to see if it happens in other editors too

Here are the links:

sample file 01 before
https://github.com/iddqd505/dotfiles/blob/main/sample%20file%2001%20before

sample file 01 before.jpg
https://github.com/iddqd505/dotfiles/blob/main/sample%20file%2001%20before.jpg

sample file 02 after
https://github.com/iddqd505/dotfiles/blob/main/sample%20file%2002%20after

sample file 02 after.jpg
https://github.com/iddqd505/dotfiles/blob/main/sample%20file%2002%20after.jpg

**Comment by chrisbra** [comment]
The mappings is probably related to patch [v9.1.0050](https://github.com/vim/vim/releases/tag/v9.1.0050). You may want to enable the experimental keycode handling for Ms-Windows, please read the related documentation at [:h :h w32-experimental-keycode-trans-strategy](https://vimhelp.org/gui_w32.txt.html#w32-experimental-keycode-trans-strategy).

**Comment by chrisbra** [comment]
I also think it looks good now. `keytrans()` correctly shows the translated keys.

**Comment by iddqd505** [comment]
These are great news thanks for the info about experimental keycode handling. I installed Vim v9.1.0067 and my Ñ mappings now work directly writting `nnoremap Ñ g$` 

However the macro register keeps changing the test file encoding from "utf-8" to "none" exactly as described in this [comment](https://github.com/vim/vim/issues/13916#issuecomment-1914683009)

I have been installing old versions of gVim and all V8.2.x I tried replicate this issue with the only difference some of them give me this error warning when opening the test file with the macro register written in line 14:
`ILLEGAL BYTE in line 14`

I had to go back to gVim version 8.1.2408 to print macro registers without "KS_EXTRA keycodes" that would change the encoding of my file but there was no keytrans() command in that version

**Comment by iddqd505** [comment]
Im out of the loop with this problem but if it helps I have found an old laptop with Windows 7 installed and an old version of gVim where I tried the test macro

`Windows 7 Ultimate Version 6.1 (Build 7601: Service Pack 1)`
`gvim82.exe 8.2.0.0 (2019 Dec 12) 32 bits`

This gVim version 8.2.0.0:
1. Doesnt print the control keycodes in the macro register paste
2. Doesnt print wrong characters in insert mode when pasting the register with `<C-r>w`
3. Doesnt have the keymap() command or help related to it
4. Doesnt change file encoding or corrupts the register paste after reopening the file

Also I found in this Windows 7 computer another gVim install file I copied there but forgot to update:
`gvim_8.2.4478_x86.exe (2022 Feb 26)`
`compiled by appveyor@APPVYR-WIN`

This updated gVim version 8.2.4478:
1. Prints the control keycodes in the macro register paste
2. Doesnt print wrong characters in insert mode when pasting the register with `<C-r>w`
3. Doesnt have the keymap() command or help related to it
4. Changes file encoding and corrupts the register paste after reopening the file

Probably not relevant but wanted to share the discovery
